### PR TITLE
feat: add admin_key to node_create.proto, node_update.proto and node.proto

### LINKED ISSUE #379: Add admin_key to node_create.proto, node_update.proto and node.proto

### PR DESCRIPTION
**Pull request by iwsimon**
**Description**:

**Related issue(s)**:

Fixes #379

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: services/node_create.proto [hunk 1]
```diff
@@ -125,4 +127,13 @@ message NodeCreateTransactionBody {
      * This field is OPTIONAL.
      */
     bytes grpc_certificate_hash = 6;
+
+    /**
+    * An administrative key for the node address book entry.
+    * <p>
+    * This key MUST sign this transaction.<br/>
+    * This key MUST sign each transaction to update this node.<br/>
+    * This field is REQUIRED and MUST NOT be set to an empty `KeyList`.
```
Comment (Neeharika-Sompalli) [inline services/node_create.proto:136]: This key must also be valid

File: services/node_update.proto [hunk 2]
```diff
@@ -148,4 +149,14 @@ message NodeUpdateTransactionBody {
      * If set, the new value SHALL replace the existing hash value.
      */
     google.protobuf.BytesValue grpc_certificate_hash = 7;
+
+    /**
+     * A new administrative key for the node address book entry.
+     * <p>
+     * If this field is set, that key MUST sign this transaction.<br/>
+     * If this field was previously set, that key MUST also sign
```
Comment (Neeharika-Sompalli) [inline services/node_update.proto:157]: This will always be set previously right since its required on `node_create`

File: services/state/addressbook/node.proto [hunk 3]
```diff
@@ -137,4 +137,12 @@ message Node {
      * node SHALL fail.
      */
     bool deleted = 9;
+
+    /**
+    * An administrative key for the node address book entry.
+    * <p>
+    * This key MUST sign each transaction to update this node.<br/>
+    * This field SHALL be set and SHALL NOT be an empty `KeyList`.
```
Comment (Neeharika-Sompalli) [inline services/state/addressbook/node.proto:145]: Here also, it should be a valid key, since we should check in code

File: services/node_create.proto (hunk 1 above)
Comment (iwsimon) [reply services/node_create.proto:136]: Updated.

File: services/node_update.proto (hunk 2 above)
Comment (iwsimon) [reply services/node_update.proto]: Updated.

File: services/node_update.proto [hunk 4]
```diff
@@ -148,4 +149,14 @@ message NodeUpdateTransactionBody {
...
+     * this transaction.<br/>
+     * This field MUST NOT be set to an empty `KeyList`.
```
[the remark this answers is not in the document]
Comment (iwsimon) [reply services/node_update.proto:158]: Updated.

File: services/state/addressbook/node.proto (hunk 3 above)
Comment (iwsimon) [reply services/state/addressbook/node.proto:145]: Updated. Thanks @Neeharika-Sompalli

File: services/node_create.proto [hunk 5]
```diff
@@ -125,4 +127,14 @@ message NodeCreateTransactionBody {
...
+    * This key MUST be valid.
```
Comment (Neeharika-Sompalli) [inline services/node_create.proto:137]: We could probably combine like this. Or any other way @jsync-swirlds  suggests

```suggestion
    /**
    * The administrative key for the node address book entry.
    * <p>
    * This field is REQUIRED. It MUST be a valid key and MUST NOT be set to an empty `KeyList`.<br/>
    * This key MUST sign this transaction and each transaction to update this node.<br/>
```

Comment (iwsimon) [reply services/node_create.proto]: Updated.

Comment (jsync-swirlds) [reply services/node_create.proto]: A couple concerns:
1. Line limit is 80 characters, this makes the lines longer (we could split the lines, but this suggestion did not).
1. The change is much less of a specification.  Specification language should be a bit more staccato than typical speech.
   * generally, replace "will" with "SHALL" to specify expectable behavior.
   * generally, each sentence is best as a separate line.
   * generally, behavior for different transactions should be separate lines.

I have a document with all the recommendations for specification text, but it's not final. I can provide that draft document separately if you're interested.

File: services/node_create.proto [hunk 6]
```diff
@@ -129,12 +129,10 @@ message NodeCreateTransactionBody {
...
-    * <p>
-    * This key MUST sign this transaction.<br/>
-    * This key MUST sign each transaction to update this node.<br/>
-    * This field is REQUIRED and MUST NOT be set to an empty `KeyList`.
-    * This key MUST be valid.
-    */
+     * The administrative key for the node address book entry.
+     * <p>
+     * This field is REQUIRED. It MUST be a valid key. Setting this to an invalid key or an empty `KeyList` will fail the transaction.<br/>
+     * This key MUST sign this transaction and each transaction to update this node.<br/>
```
Comment (jsync-swirlds) [inline services/node_create.proto:135]: The former wording works better, in my opinion.
This violates the line length for the style recommendations, and also is less specification (it reads much closer to the "documentation" style we used to write, that all needs rewriting).

File: services/node_create.proto [hunk 7]
```diff
@@ -1,6 +1,6 @@
 syntax = "proto3";
 
-package com.hedera.hapi.node.addressbook;
+package proto;
```
Comment (jsync-swirlds) [inline services/node_create.proto:3]: Please do not move these to the `proto` package.
We will not use `proto` package going forward (we never should have used it at all, as it's an example namespace not to be used in production code).

File: services/node_create.proto [hunk 8]
```diff
@@ -34,6 +34,8 @@ import "basic_types.proto";
  * to join the network, and requires governing council authorization.
  *
  * - A `NodeCreateTransactionBody` MUST be signed by the governing council.
+ % - A `NodeCreateTransactionBody` MUST be signed by the `Key` assigned to the
```
Comment (jsync-swirlds) [inline services/node_create.proto:37]: `%` ==> `*`

File: services/node_delete.proto [hunk 9]
```diff
@@ -1,6 +1,6 @@
 syntax = "proto3";
 
-package com.hedera.hapi.node.addressbook;
+package proto;
```
Comment (jsync-swirlds) [inline services/node_delete.proto:3]: Please do not move these to the `proto` package.
We will not use `proto` package going forward (we never should have used it at all, as it's an example namespace not to be used in production code).

File: services/node_update.proto [hunk 10]
```diff
@@ -1,6 +1,6 @@
 syntax = "proto3";
 
-package com.hedera.hapi.node.addressbook;
+package proto;
```
Comment (jsync-swirlds) [inline services/node_update.proto:3]: Please do not move these to the `proto` package.
We will not use `proto` package going forward (we never should have used it at all, as it's an example namespace not to be used in production code).

File: services/node_update.proto [hunk 11]
```diff
@@ -148,4 +149,12 @@ message NodeUpdateTransactionBody {
      * If set, the new value SHALL replace the existing hash value.
      */
     google.protobuf.BytesValue grpc_certificate_hash = 7;
+
+    /**
+     * The administrative key for the node address book entry.
+     * <p>
+     * This field is REQUIRED. It MUST be a valid key. Setting this to an invalid key or an empty `KeyList` will fail the transaction.<br/>
+     * This key MUST sign this transaction and each transaction to update this node.<br/>
```
Comment (jsync-swirlds) [inline services/node_update.proto:157]:
```suggestion
     * A new administrative key controlled by the node operator.
     * <p>
     * If this field is set, that key MUST sign this transaction.<br/>
     * If this field was previously set, that key MUST also sign
     * this transaction.<br/>
     * This field MUST NOT be set to an empty `KeyList`.
```

File: services/schedulable_transaction_body.proto [hunk 12]
```diff
@@ -293,16 +293,16 @@ message SchedulableTransactionBody {
     /**
      * Transaction body for a scheduled transaction to create a new node.
      */
-    com.hedera.hapi.node.addressbook.NodeCreateTransactionBody nodeCreate = 42;
```
Comment (jsync-swirlds) [inline services/schedulable_transaction_body.proto:296]: Please do not break the correct packaging (both here and below).

File: services/state/addressbook/node.proto [hunk 13]
```diff
@@ -1,6 +1,6 @@
 syntax = "proto3";
 
-package com.hedera.hapi.node.state.addressbook;
+package proto;
```
Comment (jsync-swirlds) [inline services/state/addressbook/node.proto:3]: Please do not move these to the `proto` package.
We will not use `proto` package going forward (we never should have used it at all, as it's an example namespace not to be used in production code).

File: services/state/addressbook/node.proto [hunk 14]
```diff
@@ -137,4 +137,12 @@ message Node {
      * node SHALL fail.
      */
     bool deleted = 9;
+
+    /**
+     * The administrative key for the node address book entry.
+     * <p>
+     * This field is REQUIRED. It MUST be a valid key. Setting this to an invalid key or an empty `KeyList` will fail the transaction.<br/>
+     * This key MUST sign this transaction and each transaction to update this node.<br/>
```
Comment (jsync-swirlds) [inline services/state/addressbook/node.proto:145]:
```suggestion
     * An administrative key controlled by the node operator.
     * <p>
     * This key MUST sign each transaction to update this node.<br/>
     * This field SHALL be set and SHALL NOT be an empty `KeyList`.
```

File: services/node_create.proto [hunk 15]
```diff
@@ -125,4 +127,12 @@ message NodeCreateTransactionBody {
      * This field is OPTIONAL.
      */
     bytes grpc_certificate_hash = 6;
+
+    /**
+     * The administrative key for the node address book entry.
+     * <p>
+     * This field is REQUIRED. It MUST be a valid key. Setting this to an invalid key or an empty `KeyList` will fail the transaction.<br/>
+     * This key MUST sign this transaction and each transaction to update this node.<br/>
```
Comment (jsync-swirlds) [inline services/node_create.proto:135]: The former wording works better, in my opinion.
This violates the line length for the style recommendations, and also is less specification (it reads much closer to the "documentation" style we used to write, that all needs rewriting) and more just description.
```suggestion
     * An administrative key controlled by the node operator.
     * <p>
     * This key MUST sign this transaction.<br/>
     * This key MUST sign each transaction to update this node.<br/>
     * This field is REQUIRED and MUST NOT be set to an empty `KeyList`.
```

File: services/transaction_body.proto [hunk 16]
```diff
@@ -363,22 +363,22 @@ message TransactionBody {
          * This transaction SHALL create a new consensus node record and add
          * that record to the network address book.
          */
-        com.hedera.hapi.node.addressbook.NodeCreateTransactionBody nodeCreate = 54;
+        NodeCreateTransactionBody nodeCreate = 54;
```
Comment (jsync-swirlds) [inline services/transaction_body.proto:366]: Please do not break the correct packaging (both here and below).

**Review by jsync-swirlds — CHANGES_REQUESTED** [review]
The package move is not correct.  We will _not_ be using the incorrect and improper `proto` package going forward.
A future change will move everything to proper packages, but that will also have a HIP associated because it will rewrite substantially every pre-existing file.

The rest are mostly recommendations and typos.

File: services/node_create.proto (hunk 7 above)
Comment (iwsimon) [reply services/node_create.proto:3]: The reason I had to change the package name is:
 **Key** will be invalid type if using package com.hedera.hapi.node.addressbook since it's defined in basic_types.proto and under package proto. 
 We might need to change all the package names later together.

Comment (jsync-swirlds) [reply services/node_create.proto:3]: That's not accurate.  `Key` is entirely usable, it's just referenced as `proto.Key` because of the bad package in `basic_types`.
There are several similar examples already in this file.

File: services/node_create.proto (hunk 5 above)
Comment (iwsimon) [reply services/node_create.proto]: Updated

**Comment by iwsimon** [comment]
> The package move is not correct. We will _not_ be using the incorrect and improper `proto` package going forward. A future change will move everything to proper packages, but that will also have a HIP associated because it will rewrite substantially every pre-existing file.
> 
> The rest are mostly recommendations and typos.

Updated

File: services/transaction_body.proto (hunk 16 above)
Comment (iwsimon) [reply services/transaction_body.proto]: Updated.

File: services/node_create.proto (hunk 15 above)
Comment (iwsimon) [reply services/node_create.proto]: Updated.

File: services/state/addressbook/node.proto (hunk 14 above)
Comment (iwsimon) [reply services/state/addressbook/node.proto]: Updated.

File: services/state/addressbook/node.proto (hunk 13 above)
Comment (iwsimon) [reply services/state/addressbook/node.proto]: Updated.

File: services/schedulable_transaction_body.proto (hunk 12 above)
Comment (iwsimon) [reply services/schedulable_transaction_body.proto]: Updated.

File: services/node_update.proto (hunk 11 above)
Comment (iwsimon) [reply services/node_update.proto]: Updated.

File: services/node_update.proto (hunk 10 above)
Comment (iwsimon) [reply services/node_update.proto]: Updated.

File: services/node_delete.proto (hunk 9 above)
Comment (iwsimon) [reply services/node_delete.proto]: Updated.

File: services/node_create.proto (hunk 8 above)
Comment (iwsimon) [reply services/node_create.proto]: Updated.

File: services/node_create.proto (hunk 7 above)
Comment (iwsimon) [reply services/node_create.proto]: Updated.

File: services/node_update.proto [hunk 17]
```diff
@@ -148,4 +149,14 @@ message NodeUpdateTransactionBody {
...
      */
     google.protobuf.BytesValue grpc_certificate_hash = 7;
+
+    /**
+    * An administrative key controlled by the node operator.
+     * <p>
+     * This key MUST sign this transaction.<br/>
+     * This key MUST sign each transaction to update this node.<br/>
+     * This field MUST contain a valid `Key` value.<br/>
+     * This field is REQUIRED and MUST NOT be set to an empty `KeyList`.
```
Comment (jsync-swirlds) [inline services/node_update.proto:159]: Slight error here (sorry for that).  This field is not actually _required_ in this transaction.
```suggestion
     * This field is OPTIONAL.<br/>
     * If set, this key MUST sign this transaction.<br/>
     * If set, this key MUST sign each transaction to update this node.<br/>
     * If set, this field MUST contain a valid `Key` value.<br/>
     * If set, this field MUST NOT be set to an empty `KeyList`.
```

File: services/state/addressbook/node.proto [hunk 18]
```diff
@@ -137,4 +137,14 @@ message Node {
...
      */
     bool deleted = 9;
+
+    /**
+    * An administrative key controlled by the node operator.
+     * <p>
+     * This key MUST sign this transaction.<br/>
+     * This key MUST sign each transaction to update this node.<br/>
+     * This field MUST contain a valid `Key` value.<br/>
+     * This field is REQUIRED and MUST NOT be set to an empty `KeyList`.
```
Comment (jsync-swirlds) [inline services/state/addressbook/node.proto:147]: Another small correction.
```suggestion
     * This key MUST sign each transaction to update this node.<br/>
     * This field MUST contain a valid `Key` value.<br/>
     * This field is REQUIRED and MUST NOT be set to an empty `KeyList`.
```

**Review by jsync-swirlds — APPROVED** [review]
A couple small corrections where I copy/pasted a bit too much...

File: services/state/addressbook/node.proto [hunk 19]
```diff
@@ -1,6 +1,6 @@
 syntax = "proto3";
 
-package com.hedera.hapi.node.state.addressbook;
+package com.hedera.hapi.node.addressbook;
```
Comment (jsync-swirlds) [inline services/state/addressbook/node.proto:3]: This should have the state sub-package to match what we're doing with other protobufs.

Comment (iwsimon) [reply services/state/addressbook/node.proto]: Updated.

**Review by jsync-swirlds — APPROVED** [review]

**Review by Neeharika-Sompalli — APPROVED** [review]

**Review by jsync-swirlds — APPROVED** [review]

File: services/node_create.proto [hunk 20]
```diff
@@ -125,4 +127,14 @@ message NodeCreateTransactionBody {
      * This field is OPTIONAL.
      */
     bytes grpc_certificate_hash = 6;
+
```
Comment (Neeharika-Sompalli) [inline services/node_create.proto:130]: The package name of this file is wrong `package com.hedera.hapi.node.addressbook;` It should just be `proto` similar to other services.

**Review by Neeharika-Sompalli — DISMISSED** [review]
LGTM !

File: services/node_create.proto (hunk 20 above)
Comment (jsync-swirlds) [reply services/node_create.proto:130]: As discussed in slack, `package com.hedera.hapi.node.addressbook;` is (or should be) correct.  We'll confirm that there are no issues with metrics, but barring that we should keep the more complete package directive.

**Review by mgoelswirlds — APPROVED** [review]
LGTM

**Review by Neeharika-Sompalli — DISMISSED** [review]

**Review by mgoelswirlds — APPROVED** [review]

File: services/response_code.proto [hunk 21]
```diff
@@ -1458,41 +1458,41 @@ enum ResponseCodeEnum {
...
+
+  /**
+   * ServiceEndpoint is invalid
+   */
+  INVALID_ENDPOINT = 351;
+
+  /**
+   * The number of gossip endpoints exceeds the limit
+   */
+  GOSSIP_ENDPOINTS_EXCEEDED_LIMIT = 352;
```
Comment (jsync-swirlds) [inline services/response_code.proto:1495]: Not strictly required, but might be good to fix the indentation here.
This can be a separate PR, if more efficient that way.

**Review by jsync-swirlds — DISMISSED** [review]

File: services/response_code.proto (hunk 21 above)
Comment (iwsimon) [reply services/response_code.proto]: Fixed.

**Review by Neeharika-Sompalli — APPROVED** [review]

**Review by jsync-swirlds — APPROVED** [review]
